src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.RemoteHostClientService.cs

```diff
@@ -239,10 +247,49 @@ private void OnConnectionChanged(object sender, bool connected)
                     FatalError.ReportWithoutCrash(new Exception("Connection to remote host closed"));
 
                     // use info bar to show warning to users
-                    _workspace.Services.GetService<IErrorReportingService>().ShowGlobalErrorInfo(ServicesVSResources.Unfortunately_a_process_used_by_Visual_Studio_has_encountered_an_unrecoverable_error_We_recommend_saving_your_work_and_then_closing_and_restarting_Visual_Studio,
+                    var infoBarUIs = new List<ErrorReportingUI>();
+
+                    infoBarUIs.Add(
                         new ErrorReportingUI(ServicesVSResources.Learn_more, ErrorReportingUI.UIKind.HyperLink, () =>
                             BrowserHelper.StartBrowser(new Uri(OOPKilledMoreInfoLink)), closeAfterAction: false));
+
+                    var allowRestarting = _workspace.Options.GetOption(RemoteHostOptions.RestartRemoteHostAllowed);
+                    if (allowRestarting)
+                    {
+                        infoBarUIs.Add(
+                            new ErrorReportingUI("Restart OOP", ErrorReportingUI.UIKind.Button, async () =>
+                            await RequestNewRemoteHostAsync(CancellationToken.None).ConfigureAwait(false), closeAfterAction: true));
+                    }
+
+                    _workspace.Services.GetService<IErrorReportingService>().ShowGlobalErrorInfo(
+                        ServicesVSResources.Unfortunately_a_process_used_by_Visual_Studio_has_encountered_an_unrecoverable_error_We_recommend_saving_your_work_and_then_closing_and_restarting_Visual_Studio,
+                        infoBarUIs.ToArray());
+                }
+            }
+
+            public async Task RequestNewRemoteHostAsync(CancellationToken cancellationToken)
+            {
+                var instance = await GetRemoteHostClientAsync(cancellationToken).ConfigureAwait(false);
+                if (instance == null)
+                {
+                    return;
+                }
+
+
+                // log that remote host is restarted
+                Logger.Log(FunctionId.RemoteHostClientService_Restarted, KeyValueLogMessage.NoProperty);
+
+                // we are going to kill the existing remote host, connection change is expected
+                instance.ConnectionChanged -= OnConnectionChanged;
+
+                lock (_gate)
+                {
+                    var token = _shutdownCancellationTokenSource.Token;
+                    _instanceTask = Task.Run(() => EnableAsync(token), token);
                 }
+
+                // shutdown 
+                instance.Shutdown();
```

Comment:
i find most of this code very confusing.  THere's an 'instance' but also an 'instanceTask'?  And we create the instance task here... only to shutdown the isntance?  I just don't get it and it's not clear from reading the code.